[Merged by Bors] - Retrospective invalidation of exec. payloads for opt. sync

paulhauner opened the pull request:
## Issue Addressed

NA

## Proposed Changes

Adds the functionality to allow blocks to be validated/invalidated after their import as per the [optimistic sync spec](https://github.com/ethereum/consensus-specs/blob/dev/sync/optimistic.md#how-to-optimistically-import-blocks). This means:

- Updating `ProtoArray` to allow flipping the `execution_status` of ancestors/descendants based on payload validity updates.
- Creating separation between `execution_layer` and the `beacon_chain` by creating a `PayloadStatus` struct.
- Refactoring how the `execution_layer` selects a `PayloadStatus` from the multiple statuses returned from multiple EEs.
- Adding testing framework for optimistic imports.
- Add `ExecutionBlockHash(Hash256)` new-type struct to avoid confusion between *beacon block roots* and *execution payload hashes*.
- Add `merge` to [`FORKS`](https://github.com/sigp/lighthouse/blob/c3a793fd73a3b11b130b82032904d39c952869e4/Makefile#L17) in the `Makefile` to ensure we test the beacon chain with merge settings.
    - Fix some tests here that were failing due to a missing execution layer.

## TODO

- [ ] Balance tests

paulhauner committed:
Add invalid propagation

paulhauner committed:
Adjust deltas for invalid blocks

paulhauner committed:
Avoid invalid blocks when finding the head

paulhauner committed:
Invalidate descendants of invalid blocks too

paulhauner committed:
Pass latest_valid_hash from API

paulhauner committed:
Add invalidation of payloads

paulhauner committed:
Pass root to execute_payload

paulhauner committed:
Add FixedPayloadResponse

paulhauner committed:
Add non-passing invalid payload tests

paulhauner committed:
Fix failing test

paulhauner committed:
Add simple invalidation test

paulhauner committed:
Fix errors after rebase

paulhauner committed:
Unify invalid payload processing

paulhauner committed:
Tidy, improve tests

paulhauner committed:
Add test for shutdown

paulhauner committed:
Add test for pre-finalized invalid

paulhauner committed:
Add another test

paulhauner committed:
Add handling for unknown latest valid hash

paulhauner committed:
Fix failing test, tidy

paulhauner committed:
Fix compile error in tests

ethDreamer committed:
Update Execution API to v1.0.0-alpha.6

paulhauner committed:
Add test, address bug

paulhauner committed:
Remove dbgs

paulhauner committed:
Fix failing tests by null-ing best child/desc

paulhauner committed:
Add another test

paulhauner committed:
Address clippy lint

paulhauner committed:
Remove panic from beacon chain

paulhauner committed:
Handle errors on newPayload

paulhauner committed:
Clean commented out code

paulhauner committed:
Start adding merge_sim

paulhauner committed:
Fix pattern match on exec layer

paulhauner committed:
Rename merge_sim to execution_engine_integration

paulhauner committed:
Tidy, improve testing

paulhauner committed:
Add comments, tidy

paulhauner committed:
Add to Makefile and tests

paulhauner committed:
Add to github actions

paulhauner committed:
Merge branch 'unstable' into exec-alpha6

paulhauner committed:
Remove github actions changes

paulhauner committed:
Add exec tests to github actions again

paulhauner committed:
Use serde_json instead of json

paulhauner committed:
Switch away from binary, suppress logs

paulhauner committed:
Update Makefile for tests

paulhauner committed:
Tidy

paulhauner committed:
Add `unused_port` crate

paulhauner committed:
Fix warnings and compile errors

paulhauner committed:
Merge branch 'exec-alpha6' into opt-sync

paulhauner committed:
Unify payload status parsing

paulhauner committed:
Use `PayloadStatus` in beacon chain

paulhauner committed:
Fix test compile errors

paulhauner commented:
The integration tests are currently failing on https://github.com/ethereum/go-ethereum/issues/24404

paulhauner committed:
Tidy comments around fork choice response

paulhauner committed:
Don't panic in drop

paulhauner committed:
Partially add invalidation call to fork choice

paulhauner committed:
Simplify forkchoice update method on beacon chain

paulhauner committed:
Add invalidation to fork choice

paulhauner committed:
Bump `cc` version

paulhauner commented:
I've just bumped `cc` to 1.0.73 as per https://github.com/alexcrichton/cmake-rs/issues/141.

I think this break is because Github Actions has migrated to Windows 2022 for runners.

paulhauner committed:
Merge branch 'exec-alpha6' into opt-sync

paulhauner committed:
Update geth responses

paulhauner committed:
Use windows-2019 for CI

paulhauner committed:
Use windows-2019 for CI

paulhauner committed:
Merge branch 'exec-alpha6' into opt-sync

paulhauner commented:
> The integration tests are currently failing on [ethereum/go-ethereum#24404](https://github.com/ethereum/go-ethereum/issues/24404)

That issue has been resolved.

paulhauner committed:
Skip windows tests for exec payload tests

paulhauner committed:
Update fork choice response

paulhauner committed:
Handle payload import whilst syncing in tests

paulhauner committed:
Log crit when a payload id can't be generated

paulhauner committed:
Remove unused field

paulhauner committed:
Merge branch 'exec-alpha6' into opt-sync

paulhauner committed:
Fix the mock EL FC response

paulhauner committed:
Merge branch 'unstable' into opt-sync

paulhauner commented:
I just synced this on [`merge-devnet-4`](https://notes.ethereum.org/D5xK4XrmTb6MKGudf_hcrg?view) :tada:

paulhauner committed:
Update comment

paulhauner committed:
Merge branch 'unstable' into opt-sync

paulhauner committed:
Tidy, check justified epoch

paulhauner committed:
Add ExecutionBlockHash newtype

paulhauner committed:
Fix misc test compile issues

paulhauner committed:
Ensure default payload exec status

paulhauner committed:
Add mock EL to store tests

paulhauner committed:
Fix failing test

paulhauner committed:
Un-comment release-only flag in test

paulhauner committed:
Tidy, add comment

paulhauner committed:
Check for finalized desc

paulhauner committed:
Avoid reading finalized block from DB

paulhauner committed:
Remove todo about payload ids

paulhauner committed:
Get justified root atomically

paulhauner committed:
Fix clippy lint

realbigsean commented on beacon_node/beacon_chain/src/beacon_chain.rs:
```diff
@@ -3173,6 +3177,114 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
+        // Atomically obtain the justified root from fork choice.
+        let justified_block = {
+            let fork_choice = self.fork_choice.read();
+            // De-alias 0x00..00 to the genesis block at genesis.
```
```suggestion
            // De-alias 0x00..00 to the genesis block root.
```

realbigsean commented on beacon_node/beacon_chain/src/beacon_chain.rs:
```diff
@@ -3173,6 +3177,114 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
+            fork_choice
+                .get_block(&justified_root)
+                .ok_or(Error::JustifiedMissingFromForkChoice { justified_root })?
+        };
```
maybe this should be a `get_justified_block()` method in fork choice?

paulhauner replied:
Good call, I added it here 1ed7431ba

It turns out we don't need to de-alias the genesis root. Fork choice de-aliases by default, I added some tests to ensure this here ae29684

realbigsean commented on beacon_node/beacon_chain/src/beacon_chain.rs:
```diff
@@ -3173,6 +3177,114 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
+            crit!(
+                self.log,
+                "The justified checkpoint is invalid";
+                "msg" => "ensure you are not connected to a malicious network. this error is not \
```
```suggestion
                "msg" => "ensure you are not connected to a malicious network. This error is not \
```

realbigsean commented on beacon_node/beacon_chain/src/beacon_chain.rs:
```diff
@@ -3173,6 +3177,114 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
+            )) {
+                crit!(
+                    self.log,
+                    "Unable trigger client shut down";
```
```suggestion
                    "Unable to trigger client shut down";
```

realbigsean commented on beacon_node/beacon_chain/src/beacon_chain.rs:
```diff
@@ -3188,19 +3300,51 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
+            let beacon_block_root = fork_choice.get_head(self.slot()?, &self.spec)?;
+
+            let finalized_checkpoint = fork_choice.finalized_checkpoint();
+            let finalized_block = fork_choice.get_block(&finalized_checkpoint.root).ok_or(
```
Do we have to de-alias the genesis block root here too?

paulhauner replied:
It turns out we don't need to de-alias, as per https://github.com/sigp/lighthouse/pull/2837#discussion_r815243301.

I added a `ForkChoice::get_finalized_block` over in https://github.com/sigp/lighthouse/commit/1ed7431ba2752b3bbefda99811392b138cf0d449.

realbigsean commented on beacon_node/beacon_chain/src/errors.rs:
```diff
@@ -139,15 +139,28 @@ pub enum BeaconChainError {
     BlockRewardSyncError,
     HeadMissingFromForkChoice(Hash256),
     FinalizedBlockMissingFromForkChoice(Hash256),
+    InvalidFinalizedPayload {
```
Maybe we want to include the `ExecutionBlockHash` in this error as well as the `JustifiedPayloadInvalid` variant ? Should be available when each are thrown.

paulhauner replied:
Good idea, added in 1ed7431

realbigsean commented on beacon_node/execution_layer/src/payload_status.rs:
```diff
@@ -0,0 +1,182 @@
+        .or_else(|| other_statuses.first())
+        .cloned()
+        .map(Result::Ok)
+        .unwrap_or_else(|| Err(Error::EngineErrors(errors)))
```
It seems like errors will only surface if there is no valid status? I think it could be useful to log out if there's any errors so it's easier to maintain backup execution engines

paulhauner replied:
Good call, added in 7eed122.

realbigsean commented on beacon_node/beacon_chain/src/execution_payload.rs:
```diff
@@ -57,22 +57,26 @@ pub fn notify_new_payload<T: BeaconChainTypes>(
+                Err(ExecutionPayloadError::RejectedByExecutionEngine { status }.into())
+            }
+            PayloadStatus::InvalidTerminalBlock { .. } | PayloadStatus::InvalidBlockHash { .. } => {
+                // Returning an error here should sufficient to invalidate the block. We have no
```
```suggestion
                // Returning an error here should be sufficient to invalidate the block. We have no
```

realbigsean commented on beacon_node/execution_layer/src/lib.rs:
```diff
@@ -434,7 +440,16 @@ impl ExecutionLayer {
+                                self.log(),
+                                "Exec engine unable to produce payload";
+                                "msg" => "no payload id, the engine is likely syncing. /
+                                this has potentially caused a missed block proposal.",
```
```suggestion
                                This has potentially caused a missed block proposal.",
```

paulhauner replied:
Fixed via https://github.com/sigp/lighthouse/pull/2837#discussion_r813396576

realbigsean commented on beacon_node/execution_layer/src/lib.rs:
```diff
@@ -434,7 +440,16 @@ impl ExecutionLayer {
                         .map(|response| response.payload_id)?
-                        .ok_or(ApiError::PayloadIdUnavailable)?
+                        .ok_or_else(|| {
+                            crit!(
```
Since we are potentially sending this request to multiple execution engines, this `CRIT` might be unnecessarily alarming in cases where fallback works correctly

paulhauner replied:
Good point, lowered to `error` in 5142db0

realbigsean commented on consensus/proto_array/src/proto_array.rs:
```diff
@@ -300,6 +316,180 @@ impl ProtoArray {
+    /// If `head_block_root` is not known to fork choice, an error is returned.
+    ///
+    /// If `latest_valid_hash` is `Some(hash)` where `hash` is either not known to fork choice
+    /// (perhaps it's junk orpre-finalization), then only the `head_block_root` block will be
```
```suggestion
    /// (perhaps it's junk or pre-finalization), then only the `head_block_root` block will be
```

realbigsean commented on consensus/proto_array/src/proto_array.rs:
```diff
@@ -568,6 +775,42 @@ impl ProtoArray {
     }
+
+    /// Returns `true` if the `descendant_root` has an ancestor with `ancestor_root`. Always
+    /// returns `false` if either input roots are unknown.
```
```suggestion
    /// returns `false` if either input root is unknown.
```

realbigsean commented on consensus/proto_array/src/proto_array.rs:
```diff
@@ -300,6 +316,180 @@ impl ProtoArray {
+    ///
+    /// - `Some(hash)` if the block with that payload *hash* is known to be valid and is an
+    ///     ancestor of `head_block_root`.
+    /// - `None` if the latest valid ancestor of `head_block_root` is unkown.
```
```suggestion
    /// - `None` if the latest valid ancestor of `head_block_root` is unknown.
```

realbigsean commented on consensus/proto_array/src/proto_array.rs:
```diff
@@ -300,6 +316,180 @@ impl ProtoArray {
+                | ExecutionStatus::Invalid(hash)
+                | ExecutionStatus::Unknown(hash) => {
+                    // If we're no longer processing the `head_block_root` and the last valid
+                    // ancestor is known, exit now with an error.
```
Should this be "unknown"?

```suggestion
                    // ancestor is unknown, exit now with an error.
```

realbigsean commented on consensus/proto_array/src/proto_array.rs:
```diff
@@ -300,6 +316,180 @@ impl ProtoArray {
+                        // head.
+                        node.best_child = node
+                            .best_child
+                            .filter(|best_child| invalidated_indices.contains(best_child));
```
If I'm understanding this correctly, if the best child is contained in `invalidated_indices`, this field should be set to `None`. So I think this filter condition should be inverted: `!invalidated_indices.contains(best_child)`

paulhauner replied:
Ooh yeah, good catch! :hot_pepper: I think I was being too fancy here, so I went with something simpler over in c7d0c6c.

realbigsean commented on consensus/proto_array/src/proto_array.rs:
```diff
@@ -300,6 +316,180 @@ impl ProtoArray {
+                            .best_child
+                            .filter(|best_child| invalidated_indices.contains(best_child));
+                        node.best_descendant = node.best_descendant.filter(|best_descendant| {
+                            invalidated_indices.contains(best_descendant)
```
This one too

paulhauner replied:
Also fixed in c7d0c6c

realbigsean commented on consensus/proto_array/src/proto_array.rs:
```diff
@@ -300,6 +316,180 @@ impl ProtoArray {
+                // The block is already invalid, but keep going backwards to ensure all ancestors
+                // are updated.
+                ExecutionStatus::Invalid(_) => (),
+                // This block is pre-merge, therefore it has no execution status. Nor does its
```
```suggestion
                // This block is pre-merge, therefore it has no execution status. Nor do its
```

realbigsean commented on consensus/proto_array/src/proto_array.rs:
```diff
@@ -300,6 +316,180 @@ impl ProtoArray {
+                .ok_or(Error::NodeUnknown(latest_valid_ancestor_root))?;
+            let first_potential_descendant = latest_valid_ancestor_index + 1;
+
+            // Collect all *descendants* which declared invalid since they're the descendant of a block
```
```suggestion
            // Collect all *descendants* which have been declared invalid since they're the descendant of a block
```

realbigsean commented on consensus/proto_array/src/proto_array.rs:
```diff
@@ -126,10 +129,17 @@ impl ProtoArray {
-                .ok_or(Error::InvalidNodeDelta(node_index))?;
+            let mut node_delta = if node.execution_status.is_invalid() {
+                // If the node has an invalid execution payload, reduce its weight to zero.
+                0_i64
+                    .checked_sub(node.weight as i64)
+                    .ok_or(Error::InvalidExecutionDeltaOverflow(node_index))?
```
Because we apply the proposer boost to this delta later on, it seems possible to turn this value positive again. 

That said, since we added a check for an invalid payload in `node_is_viable_for_head` and since nodes with invalid payloads shouldn't be referenced from other valid nodes I'm not sure we need to handle the delta for these nodes. They should just never be candidates for the head and have no impact on the candidacy of other nodes right?

If we want to set this to zero regardless though, maybe we should just set the `node.weight` to `0` if `node.execution_status.is_invalid()` around line 189.

paulhauner replied:
Ah yes, great point. I've addressed this over in 4e3a078

Perhaps @michaelsproul could give that commit a once-over too, since he's worked with proposer boost? :pray:

michaelsproul replied:
LGTM!

michaelsproul commented on beacon_node/execution_layer/src/lib.rs:
```diff
@@ -434,7 +440,16 @@ impl ExecutionLayer {
+                                self.log(),
+                                "Exec engine unable to produce payload";
+                                "msg" => "no payload id, the engine is likely syncing. /
+                                this has potentially caused a missed block proposal.",
```
```suggestion
                                "msg" => "No payload ID, the engine is likely syncing. \
                                          This has the potential to cause a missed block proposal.",
```

realbigsean commented on beacon_node/execution_layer/src/test_utils/mod.rs:
```diff
@@ -28,6 +27,14 @@ mod execution_block_generator;
 mod mock_execution_layer;
 
+/// Used for returning a static response for new payload calls.
+pub enum FixedPayloadResponse {
```
I'm not seeing where this is used, not sure if I'm missing it?

paulhauner replied:
Nice spotting, removed in 8da2668

realbigsean commented on beacon_node/beacon_chain/tests/payload_invalidation.rs:
```diff
@@ -0,0 +1,601 @@
+    rig.import_block(Payload::Syncing);
+}
+
+/// Ensure that an invalid payload can invalidate its parent too (give then right
```
```suggestion
/// Ensure that an invalid payload can invalidate its parent too (given the right
```

realbigsean commented on beacon_node/beacon_chain/tests/payload_invalidation.rs:
```diff
@@ -0,0 +1,601 @@
+    );
+}
+
+/// Ensure that a `latest_valid_hash` for a pre-finality block only revert a single block.
```
```suggestion
/// Ensure that a `latest_valid_hash` for a pre-finality block only reverts a single block.
```

realbigsean commented on beacon_node/beacon_chain/tests/payload_invalidation.rs:
```diff
@@ -0,0 +1,601 @@
+    // No service should have triggered a shutdown, yet.
+    assert!(rig.harness.shutdown_reasons().is_empty());
+
+    // Import a block that will invalidate the justified checkpoint.
```
```suggestion
    // Import a pre-finalized block.
```

michaelsproul commented on consensus/proto_array/src/proto_array.rs:
```diff
@@ -250,14 +260,20 @@ impl ProtoArray {
     }
 
-    pub fn propagate_execution_payload_verification(
+    /// Updates the `verified_node_index` and all descendants to have validated execution payloads.
```
```suggestion
    /// Updates the `verified_node_index` and all ancestors to have validated execution payloads.
```

michaelsproul commented on consensus/proto_array/src/proto_array.rs:
```diff
@@ -300,6 +316,180 @@ impl ProtoArray {
+                            ExecutionStatus::Unknown(hash) | ExecutionStatus::Invalid(hash) => {
+                                node.execution_status = ExecutionStatus::Invalid(*hash)
+                            }
+                            ExecutionStatus::Irrelevant(_) => (),
```
Like the `Valid` case this should be unreachable too, right? It's impossible for an invalid execution payload to have descendants with execution disabled?

paulhauner replied:
Ah yes, good point! I added an error in b3be437

michaelsproul commented on consensus/proto_array/src/proto_array.rs:
```diff
@@ -300,6 +316,180 @@ impl ProtoArray {
+    ///
+    /// If `latest_valid_hash` is `Some(hash)` where `hash` is either not known to fork choice
+    /// (perhaps it's junk orpre-finalization), then only the `head_block_root` block will be
+    /// invalidated (no ancestors). No error will be returned in this case.
```
It seems we _will_ return an error in this case: `UnknownLatestValidAncestorHash`?

We could also safely invalidate the descendents of `head_block_root` in this case too, couldn't we? Perhaps before returning the error (or instead?)

paulhauner replied:
Yep, you're totally right! In e9905aa I've removed this error and then invalidated the descendants of the head when the latest valid hash isn't available/suitable

michaelsproul commented on consensus/proto_array/src/proto_array.rs:
```diff
@@ -300,6 +316,180 @@ impl ProtoArray {
+            });
+
+        // Collect all *ancestors* which were declared invalid since they reside between the
+        // `invalid_root` and the `latest_valid_ancestor_root`.
```
```suggestion
        // `head_block_root` and the `latest_valid_ancestor_root`.
```

michaelsproul commented on consensus/proto_array/src/proto_array.rs:
```diff
@@ -300,6 +316,180 @@ impl ProtoArray {
+            }
+        }
+
+        if let Some(latest_valid_ancestor_root) = latest_valid_ancestor_root {
```
We could start from `head_block_root` here if `latest_valid_ancestor_is_descendant` is `false` (per my previous comment).

paulhauner replied:
Covered in e9905aa too :+1:

michaelsproul commented on beacon_node/beacon_chain/src/execution_payload.rs:
```diff
@@ -57,22 +57,26 @@ pub fn notify_new_payload<T: BeaconChainTypes>(
+            }
+            PayloadStatus::InvalidTerminalBlock { .. } | PayloadStatus::InvalidBlockHash { .. } => {
+                // Returning an error here should sufficient to invalidate the block. We have no
+                // information to indicate it's parent is invalid, so no need to run
```
Thinking aloud a little, we're relying on an assumption here that `notify_new_payload` is called on blocks _in order_ such that a block _must not_ have any descendants if it is found invalid at this stage.

I think that's safe, but I do wonder whether we could hit something funky in the presence of optimistic sync plus two (slashable) blocks from the same slot with different `BeaconBlockBody`s and identical `ExecutionPayload`s. The first block could be imported optimistically, with other optimistic descendants imported on top of it. Later, we import the slashable block and find that its identical payload is invalid due to the terminal block condition. We _could_ safely invalidate the descendants of the first block at this point, but will not with the code as currently written. I don't think this is a big deal however as the EE will eventually inform us of the invalidity of the descendants if we send it a fork choice update.

Also there's a typo:

```suggestion
                // information to indicate its parent is invalid, so no need to run
```

paulhauner replied:
> We could safely invalidate the descendants of the first block at this point, but will not with the code as currently written.

AIUI, the solution [in this comment](https://github.com/sigp/lighthouse/pull/2837#discussion_r815261289) has resolved this issue. Is that right?

michaelsproul replied:
I was imagining that change was already implemented and trying to think of other potential issues. But I think the thing I raised is inconsequential: we might just take one extra call to the EL to work out that we should invalidate a chain that shares an invalid payload with another chain (which is only possible if we have two distinct [slashable] blocks at the same slot). It's a non-issue :relieved:

michaelsproul commented on beacon_node/beacon_chain/src/beacon_chain.rs:
```diff
@@ -3642,31 +3754,53 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
-                        status,
-                        latest_valid_hash,
-                    })
+                    // Using a `None` latest valid ancestor will result in only the head block
+                    // being invalidated (no ancestors).
+                    self.process_invalid_execution_payload(head_block_root, None)?;
```
Flagging this as the case that I think should trigger invalidation of the block's descendants per `propagate_execution_payload_invalidation`

michaelsproul replied:
i.e. this comment https://github.com/sigp/lighthouse/pull/2837#discussion_r813467575

michaelsproul commented on beacon_node/beacon_chain/src/execution_payload.rs:
```diff
@@ -57,22 +57,26 @@ pub fn notify_new_payload<T: BeaconChainTypes>(
+            PayloadStatus::Invalid {
+                latest_valid_hash, ..
+            } => {
+                // This block has not yet been applied to fork choice, so the latest block that was
+                // imported to fork choice was the parent.
+                let latest_root = block.parent_root();
+                chain.process_invalid_execution_payload(latest_root, Some(latest_valid_hash))?;
```
Passing the parent root here seems strange because the parent is not necessarily invalid. Doing so contradicts this comment on `propagate_execution_payload_invalidation`:

> /// The `head_block_root` is the block root of the latest invalid block.

I think the logic is actually OK because `propagate_execution_payload_invalidation` gracefully handles this case, but maybe we should avoid calling it at all if `block.parent_root() == latest_valid_hash`?

paulhauner replied:
Unfortunately we can't do `block.parent_root() == latest_valid_hash` since one is a *block root* and the other a *payload hash*.

michaelsproul reviewed (changes requested):
Nice work! Another important piece for The Merge falling into place :panda_face: 

I tried to think of edge cases and break things, and will be interested to see what you think of the scenarios I mentioned. I may be paranoid/missing some things :sweat_smile: 

> Add merge to [FORKS](https://github.com/sigp/lighthouse/blob/c3a793fd73a3b11b130b82032904d39c952869e4/Makefile#L17) in the Makefile to ensure we test the beacon chain with merge settings.

I can't see this change, maybe you've got it lurking locally?

paulhauner committed:
Apply suggestions from code review

Co-authored-by: realbigsean <seananderson33@GMAIL.com>
Co-authored-by: Michael Sproul <micsproul@gmail.com>

paulhauner committed:
Add merge to makefile

paulhauner committed:
Merge branch 'unstable' into opt-sync

paulhauner committed:
Merge branch 'unstable' into opt-sync

paulhauner committed:
Add just/fin block getters to fork choice

Add just/fin block getters to fork choice

paulhauner committed:
Add execution block hashes to errors

paulhauner committed:
Update beacon_node/beacon_chain/src/execution_payload.rs

Co-authored-by: Michael Sproul <micsproul@gmail.com>

paulhauner committed:
Lower log level

paulhauner committed:
Add test for just/fin block getter

paulhauner committed:
Log errors whilst processing payload status

paulhauner committed:
Remove unused struct

paulhauner committed:
Simplify best child/desc code

paulhauner committed:
Adjust weights

paulhauner committed:
Add error for irrelevant descendant

paulhauner committed:
Invalidate more payloads

paulhauner commented:
Thanks @realbigsean and @michaelsproul! That was a great review, thanks for the improvements and bug fixes :grimacing: 

I have addressed all your comments and am waiting for CI to pass :relaxed: 

> I can't see this [`FORKS`] change, maybe you've got it lurking locally?

I must have lost this one between my desktop and laptop, thanks! Added here 2ae6d78.

michaelsproul commented on consensus/proto_array/src/proto_array.rs:
```diff
@@ -300,6 +328,204 @@ impl ProtoArray {
+                | ExecutionStatus::Unknown(hash) => {
+                    // If we're no longer processing the `head_block_root` and the last valid
+                    // ancestor is unknown, exit this loop and proceed to invalidate and
+                    // descendants of `head_block_root`.
```
```suggestion
                    // descendants of `head_block_root`/`latest_valid_ancestor_root`.
```

michaelsproul commented on consensus/proto_array/src/proto_array.rs:
```diff
@@ -300,6 +328,204 @@ impl ProtoArray {
         }
     }
 
+    /// Potentially updates blocks to have an invalid payload and therefore be ineligible for the
+    /// head.
+    ///
+    /// The `head_block_root` is the block *root* of the latest invalid block. The
+    /// `latest_valid_hash` is a payload *hash* and should be:
+    ///
+    /// - `Some(hash)` if the block with that payload *hash* is known to be valid and is an
+    ///     ancestor of `head_block_root`.
+    /// - `None` if the latest valid ancestor of `head_block_root` is unknown.
+    ///
+    /// ## Details
+    ///
+    /// If `head_block_root` is not known to fork choice, an error is returned.
+    ///
+    /// If `latest_valid_hash` is `Some(hash)` where `hash` is either not known to fork choice
+    /// (perhaps it's junk or pre-finalization), then only the `head_block_root` block will be
+    /// invalidated (no ancestors). No error will be returned in this case.
+    ///
+    /// If `latest_valid_hash` is `Some(hash)` where `hash` is a known ancestor of
+    /// `head_block_root`, then all blocks between `head_block_root` and `latest_valid_hash` will
+    /// be invalidated. Additionally, all blocks that descend from a newly-invalidated block will
+    /// also be invalidated.
```
I think this comment needs to be reworked now that the code has changed.

Specifically

> The `head_block_root` is the block *root* of the latest invalid block.

As discussed [here](https://github.com/sigp/lighthouse/pull/2837#discussion_r813545926) the block root we pass isn't necessarily the block root of an invalid block. It could be the parent of an invalid block.

Here's my attempt at an updated description:

```
Invalidate the relevant ancestors and descendants of a block with an invalid execution payload.

The `head_block_root` should be the beacon block root of the block with the invalid execution payload, _or_ its parent.

The `latest_valid_hash` should be the hash of most recent *valid* execution payload contained in an ancestor block of `head_block_root`.

This function will invalidate:

* The block matching `head_block_root` _unless_ that block has a payload matching `latest_valid_hash`.
* All ancestors of `head_block_root` back to the block with payload matching `latest_valid_hash` (endpoint > exclusive). In the case where the `head_block_root` is the parent of the invalid block and itself matches `latest_valid_hash`, no ancestors will be invalidated.
* All descendants of `latest_valid_hash` if supplied and consistent with `head_block_root`, or else all descendants of `head_block_root`.
```

paulhauner replied:
Good point! I've adapted a slightly modified version of your suggestion here: https://github.com/sigp/lighthouse/pull/2837/commits/6dee54c860eefe1bbc5a70d7ebc5353b1dc8f288. Thanks!

michaelsproul reviewed (approved):
Looking good! I'm satisfied with the latest updates, and the test coverage

Just a few docs that I think we should update, then we can merge

paulhauner committed:
Add execution_status FC tests

Start adding execution_status tests

Complete test 01

Add tests 2, 3

Appease clippy

paulhauner committed:
Update consensus/proto_array/src/proto_array.rs

Co-authored-by: Michael Sproul <micsproul@gmail.com>

paulhauner committed:
Update fork choice docs

paulhauner commented:
Thanks for the approval @michaelsproul :pray: 

I'll wait for approval/comments from @realbigsean before merging :relaxed:

paulhauner commented:
I added some non-substantive tests for the `proto_array` crate in https://github.com/sigp/lighthouse/pull/2837/commits/1ab1c433e6968f689ff036965b191adc015b3144.

I was testing to check balance propagation with and without proposer boost. I didn't find any bugs.

realbigsean reviewed (approved):
Ok changes look good to me!

paulhauner commented:
bors r+

:tada:
